primitives/src/crh/poseidon/sbox.rs

```diff
@@ -0,0 +1,256 @@
+
+            // Extract the individual inversions
+            let mut idx: i64 = w.len() as i64 - 2;
+            for i in (0..vec_state.len()).rev() {
+                for j in (0..P::T).rev() {
+                    let vec_1 = vec_state[i][j].clone();
+                    vec_state[i][j] = w_bar * &w[idx as usize];
+                    w_bar = w_bar * &vec_1;
+                    idx -= 1;
+                }
+            }
+        }
+
+        if !last {
+            // Perform the matrix mix
+            for i in 0..vec_state.len() {
+                Self::matrix_mix(&mut vec_state[i]);
+            }
+        }
+    }
+
+    fn apply_partial_batch(vec_state: &mut [Vec<F>]) {
+        let mut w: Vec<F> = Vec::new();
+        let mut accum_prod = F::one();
+
+        w.push(accum_prod);
+        // Calculate the intermediate partial products
+        for i in 0..vec_state.len() {
+            accum_prod = accum_prod * &vec_state[i][0];
+            w.push(accum_prod);
+        }
+
+        // if the accum_prod is zero, it means that one of the S-Boxes is zero
+        // in that case compute the inverses individually
+        if accum_prod == F::zero() {
+            for i in 0..(vec_state.len() - 1) {
+                if vec_state[i][0] != F::zero() {
+                    vec_state[i][0] = vec_state[i][0].inverse().unwrap();
+                }
+            }
+        } else {
+
+            // Calculate the inversion of the products
+            // Use Montgomery simultaneous inversion
+            let mut w_bar = accum_prod.inverse().unwrap();
+
+            // Extract the individual inversions
+            let mut idx: i64 = w.len() as i64 - 2;
+            for i in (0..vec_state.len()).rev() {
+                let vec_1 = vec_state[i][0].clone();
+                vec_state[i][0] = w_bar * &w[idx as usize];
+                w_bar = w_bar * &vec_1;
+                idx -= 1;
+            }
+        }
+
+        // Perform the matrix mix
+        for i in 0..vec_state.len() {
+            Self::matrix_mix(&mut vec_state[i]);
+        }
+    }
+}
+
+impl<F: PrimeField + MulShort<F, Output = F>, P: PoseidonInverseParameters<Fr = F>> PoseidonBatchSBox<P>
+for PoseidonInverseSBox<F, P> {}
+
+/// S-Box: S(x) = x^5
+#[derive(Debug)]
+pub struct PoseidonQuinticSBox<F: PrimeField, P: PoseidonParameters<Fr = F>> {
+    _field: PhantomData<F>,
+    _parameters: PhantomData<P>,
+}
+
+impl<F: PrimeField, P: PoseidonParameters<Fr = F>> PoseidonQuinticSBox<F, P> {
+    #[inline]
+    fn exp_by_5(elem: &mut F) {
+        let w1 = *elem * &(*elem);
+        let w2 = w1 * &w1;
+        *elem = w2 * &(*elem);
+    }
+}
+
+impl<F: PrimeField, P: PoseidonParameters<Fr = F>> PoseidonSBox<P> for PoseidonQuinticSBox<F, P> {
+    // It uses Montgomery multiplication
+    // Constants are defined such that the result is x * t * 2^n mod M,
+    // that is the Montgomery representation of the operand x * t mod M, and t is the 64-bit constant
```

Comment:
This comment is wrong/misleading. 
I'd just write "Uses short Montgomery multiplication".

Reply:
Done